src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h

```diff
@@ -87,4 +88,5 @@ class ControllerScriptInterfaceLegacy : public QObject {
     double getDeckRate(const QString& group);
 
     ControllerScriptEngineLegacy* m_pScriptEngineLegacy;
+    const QLoggingCategory& m_logger;
```

Comment:
Storing a reference is not recommended. Either construct a new `QLoggingCategory` or store a shared pointer. I would prefer the first option, i.e. no indirection and pointer sharing.

Unfortunately, `QLoggingCategory` is not copyable.

Reply:
I presumed that it does not work to create multiple QLoggingCategories that share the same category string because it is not copyable, but I have not actually tried. std::shared_ptr could work though.

Reply:
I'll use a shared pointer to avoid having to pass a `const char*` around...

Reply:
Idea: Create a copyable wrapper around QLoggingCategory that internally stores both the category as a QByteArray and the QLoggingCategory as a shared_pointer. It should provide an implicit conversion operator to `const QLoggingCategory&`.

@daschuer Or are instances of QLoggingCategory supposed to be created *on-the-fly* instead of copying them?

Reply:
The wrapper could be named `DynamicLoggingCategory`.

The shared pointer is probably not needed. You simply instantiate a new QLoggingCategory instance from the category in the QByteArray member.

Reply:
My understanding is that QLoggingCategory represents just an entry in the logger registry. Multiple instances share the same entry -> *Flyweight pattern*. Unfortunately, this is not documented properly.

Reply:
> I presumed that it does not work to create multiple QLoggingCategories that share the same category string because it is not copyable

I was wrong. That does work: https://github.com/Be-ing/mixxx/commit/50f659cecb55c68df66e0037dda694500be2015d

Reply:
I think it could work to pass `QByteArray(m_logBaseRawString.constData())` between classes, but then classes that need to pass the string onto another class they create would need both a QByteArray and QLoggingCategory member... ugly... :/

Reply:
Oh, [QLoggingCategory::categoryName](https://doc.qt.io/qt-5/qloggingcategory.html#categoryName) returns the `const char*`, so no need to store a QByteArray in each class. I think using `QByteArray(m_logBase.categoryName())` to make a deep copy of the string to pass between classes would work.

Reply:
I sure wish specifically defined lifetimes checked by the compiler were part of the language right now... :crab:

Reply:
Alternatively we could just pass the `const char*` around directly without making copies. The lifetimes of all the objects which passed the `const char*` are within the lifetimes of the Controller so that would work, but it could be dangerous with future refactoring. I don't have a strong opinion, what do you think?

Reply:
Or we could pass a plain `QByteArray` around and rely on Qt's implicit sharing

Reply:
No hacks, please. Use a wrapper with an internal QByteArray for dynamic logging categories.

Reply:
I implemented a wrapper. The convenience comes at the cost of [grabbing a mutex](https://code.woboq.org/qt5/qtbase/src/corelib/io/qloggingregistry.cpp.html#_ZN16QLoggingRegistry16registerCategoryEP16QLoggingCategory9QtMsgType) with each implicit conversion to to QLoggingCategory. I think that cost is worth it considering logging is too slow for real time code anyway.

Reply:
Ouch, I should have looked into the implementation of QLoggingCategory. Then using a `smart_ptr` is probably better.

We need a `SharedLoggingCategory` that is both copyable and movable. Internally it stores a `smart_ptr` to an immutable, non-copyable/non-movable `DynamicLoggingCategory` (all members const, copy/move ctor deleted) which contains both the category string as a `QByteArray` and the actual `QLoggingCategory`.

Could this work?

Reply:
I found a simpler solution. I gave DynamicLoggingCategory a QLoggingCategory member and made the `operator()()` return a const reference to the member instead of creating a new QLoggingCategory for each call. I also realized I could make the members private and the copy constructor can still access the QByteArray.